File: internal/validate/validate.go
// Package validate implements `decks validate`: an independent, spec-driven
// checker for deck.html files (spec claude-code-integration "the sleeper feature", spec principles-and-invariants validation
// + offline-first).
//
// WHY a separate Go validator (it does NOT reuse the TS model):
//
// The editor's byte-stable parse→model→serialize round-trip lives in the
// TypeScript model (web/src/lib/model). That code is the source of truth for
// *editing*. This Go package deliberately re-implements the SAME spec rules
// independently so that decks authored OUTSIDE the editor — by Claude Code or a
// human hand-editing HTML — can be gated before they reach the canvas
// (spec claude-code-integration: "Run by both Claude Code and the editor's save path, so malformed
// decks are caught instead of silently breaking the canvas"). Two independent
// implementations of one contract is intentional: it catches drift in either.
//
// The checks (spec claude-code-integration / principles-and-invariants / layout-vocabulary):
//
//	(a) data-lay / data-align / data-justify and numeric data-* values are in
//	    the ALLOWED sets (the layout contract, spec layout-vocabulary).
//	(b) data-eid uniqueness (no duplicates — stable ids must be unique to target).
//	(c) asset existence: referenced relative srcs/hrefs resolve to files that
//	    exist under the deck folder (excluding http(s):// and data: URLs).
//	(d) parse / well-formedness: the HTML parses and tags are balanced enough to
//	    be safe to render. We use golang.org/x/net/html's tokenizer to detect
//	    gross malformation (mismatched / unclosed / stray tags, tokenizer errors).
//	(e) offline guard (cross-cutting X-1, spec principles-and-invariants): ZERO external http(s):// (or
//	    protocol-relative //) resource URLs — every dependency must be vendored.
package validate

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"golang.org/x/net/html"
)

// Issue is a single validation diagnostic. Line and EID are best-effort context
// (omitted from JSON when zero/empty) so callers can point the author at the
// offending element.
type Issue struct {
	Code    string `json:"code"`
	Message string `json:"message"`
	Line    int    `json:"line,omitempty"`
	EID     string `json:"eid,omitempty"`
}

// Result is the outcome of validating a deck. OK is true only when Errors is
// empty. It marshals directly to the HTTP endpoint's JSON contract:
//
//	{"ok":bool,"errors":[{"code","message","line"?,"eid"?}]}
type Result struct {
	OK     bool    `json:"ok"`
	Errors []Issue `json:"errors"`
}

// Allowed enum sets — the layout contract (spec layout-vocabulary). These mirror the TS
// LayValue / AlignValue / JustifyValue unions in web/src/lib/model/layout.ts,
// kept in sync by hand (see package doc on why duplication is intentional).
//
// allowedTheme mirrors THEME_NAMES in web/src/lib/model/theme.ts (P10).
// Both lists must stay identical — they are independent re-implementations of
// the same spec contract (same reason as layout: catch drift in either).
var (
	allowedLay     = map[string]bool{"stack": true, "row": true, "grid": true, "layers": true}
	allowedAlign   = map[string]bool{"start": true, "center": true, "end": true, "stretch": true}
	allowedJustify = map[string]bool{"start": true, "center": true, "end": true, "between": true, "around": true}
	// allowedTheme — bundled reveal.js theme names (spec P10). Keep in sync with
	// THEME_NAMES in web/src/lib/model/theme.ts.
	allowedTheme = map[string]bool{
		"black": true, "white": true, "league": true, "beige": true,
		"night": true, "moon": true, "solarized": true, "solarized-dark": true,
		"dracula": true, "sky": true,
	}
)

// voidElements never have an end tag (HTML5 void elements). They are skipped by
// the tag-balance check so a self-closing <img>/<br>/<meta> is not flagged as
// "unclosed".
var voidElements = map[string]bool{
	"area": true, "base": true, "br": true, "col": true, "embed": true,
	"hr": true, "img": true, "input": true, "link": true, "meta": true,
	"param": true, "source": true, "track": true, "wbr": true,
}

// optionalEndElements may legally omit their end tag in HTML5 (the parser infers
// it). We exclude them from strict balance checking so valid markup like
// "<li>a<li>b" or "<td>x<td>y" is not reported as malformed. Their content is
// still tokenized; we simply do not require a matching close.
var optionalEndElements = map[string]bool{
	"p": true, "li": true, "dt": true, "dd": true, "td": true, "th": true,
	"tr": true, "thead": true, "tbody": true, "tfoot": true, "option": true,
	"optgroup": true, "colgroup": true, "rp": true, "rt": true,
}

// resourceAttr maps an element name to the attribute that carries a resource
// URL we must (a) verify exists on disk when relative and (b) reject when
// external. Only attributes that reference loadable assets are listed; <a href>
// is intentionally excluded (it is navigation, commonly a "#" fragment).
var resourceAttr = map[string]string{
	"img":    "src",
	"script": "src",
	"link":   "href",
	"video":  "src",
	"audio":  "src",
	"source": "src",
	"track":  "src",
	"iframe": "src",
	"embed":  "src",
}

// Deck validates decks/<name>'s deck.html on disk. deckDir is the deck folder
// (…/decks/<name>); relative asset references are resolved against it.
func Deck(deckDir string) (Result, error) {
	src, err := os.ReadFile(filepath.Join(deckDir, "deck.html"))
	if err != nil {
		return Result{}, fmt.Errorf("validate: read deck.html: %w", err)
	}
	return Bytes(src, deckDir), nil
}

// Bytes validates raw deck HTML. deckDir is used only to resolve relative asset
// references for the existence check; pass "" to skip asset existence (the enum,
// eid, well-formedness, and external-URL checks still run).
//
// All checks run in a SINGLE tokenizer pass so every Issue can carry a line
// number. The x/net/html tokenizer is a streaming HTML5 tokenizer: it never
// "fixes up" the tree the way html.Parse does, which is exactly what we want —
// we are validating the author's literal bytes, not a normalized DOM.
func Bytes(src []byte, deckDir string) Result {
	var issues []Issue

	z := html.NewTokenizer(strings.NewReader(string(src)))

	// line tracks the 1-based starting line of the current token by counting
	// newlines consumed in prior tokens' raw bytes.
	line := 1
	consumed := 0

	// seenEIDs maps a data-eid to the line it first appeared on, so a duplicate
	// can name where the collision is.
	seenEIDs := map[string]int{}

	// openStack is the balance stack of "must-close" elements (non-void,
	// non-optional). Each entry remembers the tag and its line for diagnostics.
	type openTag struct {
		name string
		line int
	}
	var openStack []openTag

tokenize:
	for {
		tt := z.Next()
		// The starting line of this token is 1 + all newlines consumed before it.
		line = 1 + consumed
		raw := z.Raw()
		consumed += strings.Count(string(raw), "\n")

		switch tt {
		case html.ErrorToken:
			err := z.Err()
			if err == io.EOF {
				break tokenize
			}
			// A non-EOF tokenizer error means the bytes could not be tokenized at
			// all — gross malformation (spec claude-code-integration (d)).
			issues = append(issues, Issue{
				Code:    "malformed-html",
				Message: "HTML could not be tokenized: " + err.Error(),
				Line:    line,
			})
			break tokenize

		case html.StartTagToken, html.SelfClosingTagToken:
			nameBytes, hasAttr := z.TagName()
			name := strings.ToLower(string(nameBytes))

			// Collect attributes into a map for this element (last value wins,
			// matching browser behaviour).
			attrs := map[string]string{}
			if hasAttr {
				for {
					k, v, more := z.TagAttr()
					attrs[strings.ToLower(string(k))] = string(v)
					if !more {
						break
					}
				}
			}

			issues = append(issues, checkElement(name, attrs, line, seenEIDs, deckDir)...)

			// Tag-balance bookkeeping (well-formedness, spec claude-code-integration (d)).
			// SelfClosingTagToken (<x/>) and void elements never open a scope.
			if tt == html.StartTagToken && !voidElements[name] && !optionalEndElements[name] {
				openStack = append(openStack, openTag{name: name, line: line})
			}

		case html.EndTagToken:
			nameBytes, _ := z.TagName()
			name := strings.ToLower(string(nameBytes))
			if voidElements[name] || optionalEndElements[name] {
				continue // not tracked on the stack
			}
			if len(openStack) == 0 {
				issues = append(issues, Issue{
					Code:    "malformed-html",
					Message: fmt.Sprintf("stray closing tag </%s> with no matching open tag", name),
					Line:    line,
				})
				continue
			}
			top := openStack[len(openStack)-1]
			if top.name != name {
				// Mismatched nesting (e.g. <div><span></div>) — proper nesting is
				// required for the tracked element set.
				issues = append(issues, Issue{
					Code:    "malformed-html",
					Message: fmt.Sprintf("mismatched closing tag </%s>: expected </%s> (opened at line %d)", name, top.name, top.line),
					Line:    line,
				})
				continue
			}
			openStack = openStack[:len(openStack)-1]
		}
	}

	// Anything left open at EOF was never closed.
	for _, o := range openStack {
		issues = append(issues, Issue{
			Code:    "malformed-html",
			Message: fmt.Sprintf("unclosed <%s> tag", o.name),
			Line:    o.line,
		})
	}

	return Result{OK: len(issues) == 0, Errors: issues}
}

// checkElement runs the per-element spec checks (enums, numerics, eid
// uniqueness, asset existence, external URLs) and returns any issues.
func checkElement(name string, attrs map[string]string, line int, seenEIDs map[string]int, deckDir string) []Issue {
	var issues []Issue
	eid := attrs["data-eid"]

	// (b) data-eid uniqueness.
	if eid != "" {
		if first, dup := seenEIDs[eid]; dup {
			issues = append(issues, Issue{
				Code:    "duplicate-eid",
				Message: fmt.Sprintf("duplicate data-eid %q (first seen at line %d)", eid, first),
				Line:    line,
				EID:     eid,
			})
		} else {
			seenEIDs[eid] = line
		}
	}

	// (a) enum attributes.
	if v, ok := attrs["data-lay"]; ok && !allowedLay[v] {
		issues = append(issues, enumIssue("data-lay", v, "stack|row|grid|layers", line, eid))
	}
	if v, ok := attrs["data-align"]; ok && !allowedAlign[v] {
		issues = append(issues, enumIssue("data-align", v, "start|center|end|stretch", line, eid))
	}
	if v, ok := attrs["data-justify"]; ok && !allowedJustify[v] {
		issues = append(issues, enumIssue("data-justify", v, "start|center|end|between|around", line, eid))
	}
	if v, ok := attrs["data-visibility"]; ok && v != "hidden" {
		// Only data-visibility="hidden" is defined (the slide-hide flag, layout
		// contract). Any other value is a typo that would silently do nothing.
		issues = append(issues, enumIssue("data-visibility", v, "hidden", line, eid))
	}
	// data-theme is only meaningful on <section> elements (P10). Any value not in
	// the bundled theme list is flagged. data-background-color is a reveal.js
	// pass-through attribute and is always tolerated (no validation). Inline
	// --r-* CSS custom properties in style= are also silently tolerated so that
	// P10-3/4 output passes validation without changes here.
	if v, ok := attrs["data-theme"]; ok && name == "section" && !allowedTheme[v] {
		issues = append(issues, enumIssue("data-theme", v,
			"black|white|league|beige|night|moon|solarized|solarized-dark|dracula|sky", line, eid))
	}

	// data-background-* set (reveal.js per-slide background, spec theming-and-styles). ALL of
	// these are TOLERATED with no enum: -color, -size, -position, -repeat,
	// -opacity, -gradient, -video-loop, -video-muted are plain pass-through
	// strings reveal.js renders natively in 5.x. The two that reference loadable
	// assets — data-background-image and data-background-video — get the same
	// treatment as <img src>/<video src>: the offline X-1 guard flags http(s)://
	// values, and LOCAL relative paths are asset-existence-checked. Everything
	// else passes through untouched.
	//
	// Dual-encoded: also see getThemeProps/setThemeProps + BACKGROUND_ATTRS in
	// web/src/lib/model/theme.ts (keep the attribute set in sync). The TS model
	// round-trips the whole set; this validator only constrains the two asset
	// references and the offline guard.
	for _, bgAsset := range []string{"data-background-image", "data-background-video"} {
		if v, ok := attrs[bgAsset]; ok {
			issues = append(issues, checkResourceURL(name, bgAsset, v, line, eid, deckDir)...)
		}
	}

	// data-layout (P14): layout-preset MARKER on <section> elements.
	// Intentionally NOT enum-restricted — the value is the preset name (e.g.
	// "title-body", "two-column") chosen from the preset list, but that list is
	// open-ended and must not be encoded here to avoid coupling. Any non-empty
	// string is valid; the marker has no reflow semantics of its own.
	// Dual-encoded: also see getLayoutMarker/setLayoutMarker in
	// web/src/lib/model/layout.ts (keep in sync on attribute name).
	if v, ok := attrs["data-layout"]; ok && name == "section" && strings.TrimSpace(v) == "" {
		issues = append(issues, Issue{
			Code:    "invalid-attr",
			Message: "data-layout on <section> must be a non-empty string (preset name)",
			Line:    line,
			EID:     eid,
		})
	}

	// data-slot (P14): named-slot MARKER on any element.
	// Identifies the semantic role of a container within a preset layout (e.g.
	// "content", "sidebar"). Any non-empty string is valid; the attribute has no
	// reflow semantics of its own — it is purely informational for preset tooling.
	// Dual-encoded: also see getSlot/setSlot in web/src/lib/model/layout.ts
	// (keep in sync on attribute name).
	if v, ok := attrs["data-slot"]; ok && strings.TrimSpace(v) == "" {
		issues = append(issues, Issue{
			Code:    "invalid-attr",
			Message: "data-slot must be a non-empty string (slot name)",
			Line:    line,
			EID:     eid,
		})
	}

	// data-chart / data-chart-data (P17-15): Chart.js chart block on a <canvas>.
	//   • data-chart      — the chart TYPE string ("bar", "line", "pie", …). Any
	//                       non-empty string is valid (the Chart.js type list is
	//                       open-ended; do not enum-couple it here). Empty → error.
	//   • data-chart-data — a JSON Chart.js config `{type, data, options?}` the
	//                       runtime plugin JSON.parses. It MUST be present and
	//                       parse as JSON; malformed JSON is flagged invalid-attr.
	// Dual-encoded: also see getChartProps/setChartProps in
	// web/src/lib/model/layout.ts and the canvas-data-chart leaf rule in
	// web/src/lib/model/classify.ts (keep in sync on the attribute names).
	if v, ok := attrs["data-chart"]; ok {
		if strings.TrimSpace(v) == "" {
			issues = append(issues, Issue{
				Code:    "invalid-attr",
				Message: "data-chart must be a non-empty string (chart type)",
				Line:    line,
				EID:     eid,
			})
		}
		// A chart marker implies a data config; require it to be present + parseable.
		raw, hasData := attrs["data-chart-data"]
		if !hasData {
			issues = append(issues, Issue{
				Code:    "invalid-attr",
				Message: "data-chart requires a data-chart-data JSON config attribute",
				Line:    line,
				EID:     eid,
			})
		} else if !json.Valid([]byte(raw)) {
			issues = append(issues, Issue{
				Code:    "invalid-attr",
				Message: "data-chart-data must be parseable JSON (Chart.js config)",
				Line:    line,
				EID:     eid,
			})
		}
	} else if raw, ok := attrs["data-chart-data"]; ok && !json.Valid([]byte(raw)) {
		// data-chart-data present without data-chart (unusual) — still validate it.
		issues = append(issues, Issue{
			Code:    "invalid-attr",
			Message: "data-chart-data must be parseable JSON (Chart.js config)",
			Line:    line,
			EID:     eid,
		})
	}
	// data-qr / data-qr-* (P19): QR code block on a <div>.
	//   • data-qr       — the encoded PAYLOAD (URL/text). Any non-empty string is
	//                      valid; empty → error (nothing to encode).
	//   • data-qr-ec    — error-correction level, one of L | M | Q | H. Other
	//                      values are rejected (the generator only accepts these).
	//   • data-qr-quiet — quiet-zone width in modules, a non-negative integer
	//                      (validated with the numeric attributes below).
	//   • data-qr-fg / data-qr-bg — free-form colour strings (no constraint here).
	// Dual-encoded: also see getQrProps/setQrProps in web/src/lib/model/layout.ts
	// and the div-data-qr leaf rule in web/src/lib/model/classify.ts (keep in sync
	// on the attribute names and the EC enum).
	if v, ok := attrs["data-qr"]; ok && strings.TrimSpace(v) == "" {
		issues = append(issues, Issue{
			Code:    "invalid-attr",
			Message: "data-qr must be a non-empty string (the encoded payload)",
			Line:    line,
			EID:     eid,
		})
	}
	if v, ok := attrs["data-qr-ec"]; ok {
		switch strings.TrimSpace(v) {
		case "L", "M", "Q", "H":
			// valid
		default:
			issues = append(issues, Issue{
				Code:    "invalid-attr",
				Message: fmt.Sprintf("data-qr-ec=%q must be one of L, M, Q, H", v),
				Line:    line,
				EID:     eid,
			})
		}
	}

	// data-footer-hidden (P17-18): boolean MARKER on a <section> opting that slide
	// out of the deck-level footer overlay (a managed custom.css rule keyed off
	// section:not([data-footer-hidden])). It is presence-only — any value (incl.
	// the empty boolean-attribute form) is tolerated, so there is nothing to
	// reject here; the clause documents the recognised attribute and keeps the
	// contract dual-encoded with getFooterHidden/setFooterHidden in
	// web/src/lib/model/layout.ts (keep in sync on the attribute name).
	_ = attrs["data-footer-hidden"]

	// (a) numeric attributes. gap/pad/grow are non-negative integers; span is a
	// positive integer (>=1). These are the documented integer attributes of the
	// layout contract.
	for _, n := range []struct {
		attr string
		min  int
	}{
		{"data-gap", 0}, {"data-pad", 0}, {"data-grow", 0}, {"data-span", 1},
		// data-autoslide (P17-20): reveal-native per-slide auto-advance, in ms.
		// Non-negative integer (0 = pause on this slide). Only emitted on
		// <section> by the editor, but validated wherever present.
		{"data-autoslide", 0},
		// data-qr-quiet (P19): QR quiet-zone width in modules. Non-negative int.
		{"data-qr-quiet", 0},
	} {
		if v, ok := attrs[n.attr]; ok {
			if iv, err := strconv.Atoi(strings.TrimSpace(v)); err != nil || iv < n.min {
				issues = append(issues, Issue{
					Code:    "invalid-numeric",
					Message: fmt.Sprintf("%s=%q must be an integer >= %d", n.attr, v, n.min),
					Line:    line,
					EID:     eid,
				})
			}
		}
	}

	// Free-element coordinates: numeric (floats allowed); w/h must be >= 0.
	for _, n := range []struct {
		attr   string
		nonNeg bool
	}{
		{"data-x", false}, {"data-y", false}, {"data-rot", false},
		{"data-w", true}, {"data-h", true},
	} {
		if v, ok := attrs[n.attr]; ok {
			fv, err := strconv.ParseFloat(strings.TrimSpace(v), 64)
			if err != nil || (n.nonNeg && fv < 0) {
				bound := "a number"
				if n.nonNeg {
					bound = "a number >= 0"
				}
				issues = append(issues, Issue{
					Code:    "invalid-numeric",
					Message: fmt.Sprintf("%s=%q must be %s", n.attr, v, bound),
					Line:    line,
					EID:     eid,
				})
			}
		}
	}

	// (c)/(e) resource URLs: external-URL guard + asset existence.
	if attr, ok := resourceAttr[name]; ok {
		if url, present := attrs[attr]; present {
			issues = append(issues, checkResourceURL(name, attr, url, line, eid, deckDir)...)
		}
	}

	return issues
}

// enumIssue builds an "invalid-enum" diagnostic.
func enumIssue(attr, got, allowed string, line int, eid string) Issue {
	return Issue{
		Code:    "invalid-enum",
		Message: fmt.Sprintf("%s=%q is not allowed (expected one of %s)", attr, got, allowed),
		Line:    line,
		EID:     eid,
	}
}

// checkResourceURL enforces the offline guard (no external URLs) and asset
// existence for a single resource reference.
func checkResourceURL(tag, attr, url string, line int, eid, deckDir string) []Issue {
	raw := strings.TrimSpace(url)
	if raw == "" {
		return nil
	}

	lower := strings.ToLower(raw)

	// (e) Offline guard (X-1, spec principles-and-invariants): zero external resource URLs. http(s)://
	// and protocol-relative // both reach the network.
	if strings.HasPrefix(lower, "http://") || strings.HasPrefix(lower, "https://") || strings.HasPrefix(raw, "//") {
		return []Issue{{
			Code:    "external-url",
			Message: fmt.Sprintf("<%s %s=%q> is an external URL; decks must be offline-first (vendor it locally, spec principles-and-invariants)", tag, attr, url),
			Line:    line,
			EID:     eid,
		}}
	}

	// Non-file schemes that are legitimately offline/inline: skip existence.
	if strings.HasPrefix(lower, "data:") || strings.HasPrefix(lower, "mailto:") ||
		strings.HasPrefix(lower, "tel:") || strings.HasPrefix(raw, "#") {
		return nil
	}

	// (c) Asset existence: only meaningful when we know the deck folder.
	if deckDir == "" {
		return nil
	}

	// Strip query/fragment so "img/x.png?v=2#a" resolves to the file.
	path := raw
	if i := strings.IndexAny(path, "?#"); i >= 0 {
		path = path[:i]
	}
	path = filepath.FromSlash(path)

	// Resolve relative to the deck folder and confine to it: a reference that
	// escapes the deck (../../etc/passwd) is itself an error — decks are
	// self-contained (spec principles-and-invariants).
	resolved := filepath.Join(deckDir, path)
	rel, err := filepath.Rel(deckDir, resolved)
	if err != nil || rel == ".." || strings.HasPrefix(rel, ".."+string(filepath.Separator)) {
		return []Issue{{
			Code:    "asset-escapes-deck",
			Message: fmt.Sprintf("<%s %s=%q> resolves outside the deck folder; decks must be self-contained", tag, attr, url),
			Line:    line,
			EID:     eid,
		}}
	}

	if _, err := os.Stat(resolved); err != nil {
		return []Issue{{
			Code:    "missing-asset",
			Message: fmt.Sprintf("<%s %s=%q> references a file that does not exist in the deck", tag, attr, url),
			Line:    line,
			EID:     eid,
		}}
	}
	return nil
}

File: web/src/lib/model/validate.ts
/**
 * validate.ts — Client-side deck validation (P8-3 / spec claude-code-integration "slides validate").
 *
 * WHY THIS EXISTS:
 * ================
 * Both Claude Code and the editor's save path write `deck.html`; spec claude-code-integration says a
 * malformed deck must be "caught instead of silently breaking the canvas". The
 * single source of truth for validation is the Go `slides validate` /
 * POST /api/decks/{name}/validate endpoint, but the editor ALSO keeps this fast,
 * synchronous, offline guard so:
 *
 *   • a save that would break the model is surfaced even when the backend
 *     endpoint is unavailable, and
 *   • we never have to round-trip the network to know the obvious cases
 *     (unparseable HTML, duplicate eids, illegal layout-contract values).
 *
 * This mirrors the layout-contract checks already enforced (by THROWING) in
 * layout.ts, but here they are NON-throwing: we collect every problem so the UI
 * can show the user the full list and let THEM decide, rather than aborting on
 * the first error (spec claude-code-integration "show the errors and let the user decide").
 *
 * PURE: parse + tree walk only, no DOM, fully unit-testable (validate.test.ts).
 */

import { parseDeck } from './parse';
import { serializeDeck } from './serialize';
import { walk, getAttribute } from './edit';
import type { DeckModel, ElementNode } from './types';

/** A single validation problem. `code` is machine-stable; `message` is for humans. */
export interface ValidationError {
  /** Stable machine code, e.g. 'parse' | 'round-trip' | 'duplicate-eid' | 'bad-lay'. */
  code: string;
  /** Human-readable description shown in the validation banner. */
  message: string;
  /** The offending element's `data-eid`, when one can be attributed. */
  eid?: string;
}

/** The outcome of validating a deck. `ok === (errors.length === 0)`. */
export interface ValidationResult {
  ok: boolean;
  errors: ValidationError[];
}

// ─── Layout-contract allowed value sets (spec layout-vocabulary) ───────────────────────────

const LAY_VALUES = new Set(['stack', 'row', 'grid', 'layers']);
const ALIGN_VALUES = new Set(['start', 'center', 'end', 'stretch']);
const JUSTIFY_VALUES = new Set(['start', 'center', 'end', 'between', 'around']);

/** Match a clean signed integer string ("0", "24", "-3"). */
const INT_RE = /^-?\d+$/;
/** Match a clean (optionally signed/decimal) number string for free coords. */
const NUM_RE = /^-?\d+(?:\.\d+)?$/;

/** Parse an integer attribute; returns null if the raw string is not a clean int. */
function asInt(raw: string): number | null {
  return INT_RE.test(raw.trim()) ? parseInt(raw, 10) : null;
}

/**
 * Validate the layout contract (spec layout-vocabulary) across a parsed model: enum values are
 * legal, numeric attributes are well-formed and in-range, and every `data-eid`
 * is unique. Returns ALL problems found (does not stop at the first).
 *
 * Exposed separately from {@link validateSource} so tests can feed a model
 * directly and so callers that already hold a model avoid re-parsing.
 */
export function validateModel(model: DeckModel): ValidationResult {
  const errors: ValidationError[] = [];
  const seenEids = new Map<string, number>();

  walk(model, (node) => {
    if (node.type !== 'element') return;
    const el = node as ElementNode;
    const eid = getAttribute(el, 'data-eid') ?? undefined;

    // ── Unique data-eid (spec document-model): a duplicate breaks selection + Claude targeting.
    if (eid) seenEids.set(eid, (seenEids.get(eid) ?? 0) + 1);

    // ── Enum attributes (spec layout-vocabulary LAYOUT CONTRACT) ──────────────────────────
    checkEnum(errors, el, eid, 'data-lay', LAY_VALUES);
    checkEnum(errors, el, eid, 'data-align', ALIGN_VALUES);
    checkEnum(errors, el, eid, 'data-justify', JUSTIFY_VALUES);

    // ── Non-negative integers: gap / pad / grow ────────────────────────────
    checkNonNegInt(errors, el, eid, 'data-gap');
    checkNonNegInt(errors, el, eid, 'data-pad');
    checkNonNegInt(errors, el, eid, 'data-grow');

    // ── Positive integer: span (>= 1) ──────────────────────────────────────
    const spanRaw = getAttribute(el, 'data-span');
    if (spanRaw !== null) {
      const n = asInt(spanRaw);
      if (n === null || n < 1) {
        errors.push({
          code: 'bad-span',
          message: `data-span must be an integer >= 1 (got "${spanRaw}")`,
          eid,
        });
      }
    }

    // ── Free coordinates: x/y any number; w/h non-negative numbers ──────────
    checkNumber(errors, el, eid, 'data-x', false);
    checkNumber(errors, el, eid, 'data-y', false);
    checkNumber(errors, el, eid, 'data-w', true);
    checkNumber(errors, el, eid, 'data-h', true);
  });

  // Flag every eid used more than once (report once per duplicated id).
  for (const [eid, count] of seenEids) {
    if (count > 1) {
      errors.push({
        code: 'duplicate-eid',
        message: `data-eid "${eid}" is used ${count} times; eids must be unique`,
        eid,
      });
    }
  }

  return { ok: errors.length === 0, errors };
}

function checkEnum(
  errors: ValidationError[],
  el: ElementNode,
  eid: string | undefined,
  attr: string,
  allowed: Set<string>,
): void {
  const raw = getAttribute(el, attr);
  if (raw !== null && !allowed.has(raw)) {
    errors.push({
      code: `bad-${attr.replace('data-', '')}`,
      message: `${attr} must be one of {${[...allowed].join(', ')}} (got "${raw}")`,
      eid,
    });
  }
}

function checkNonNegInt(
  errors: ValidationError[],
  el: ElementNode,
  eid: string | undefined,
  attr: string,
): void {
  const raw = getAttribute(el, attr);
  if (raw === null) return;
  const n = asInt(raw);
  if (n === null || n < 0) {
    errors.push({
      code: `bad-${attr.replace('data-', '')}`,
      message: `${attr} must be a non-negative integer (got "${raw}")`,
      eid,
    });
  }
}

function checkNumber(
  errors: ValidationError[],
  el: ElementNode,
  eid: string | undefined,
  attr: string,
  nonNegative: boolean,
): void {
  const raw = getAttribute(el, attr);
  if (raw === null) return;
  const ok = NUM_RE.test(raw.trim());
  const n = ok ? parseFloat(raw) : NaN;
  if (!ok || (nonNegative && n < 0)) {
    errors.push({
      code: `bad-${attr.replace('data-', '')}`,
      message: `${attr} must be a${nonNegative ? ' non-negative' : ''} number (got "${raw}")`,
      eid,
    });
  }
}

/**
 * Validate a raw HTML source string end-to-end (the editor's save guard):
 *
 *   1. PARSE — it must parse without throwing.
 *   2. ROUND-TRIP IDEMPOTENCY — serialize(model) then re-parse+serialize must be
 *      byte-identical. If they diverge, persisting `source` would silently alter
 *      the document on the next load (a "save that would break the model"), so
 *      we surface it (spec document-model idempotent-round-trip invariant; spec claude-code-integration "HTML
 *      parses and round-trips").
 *   3. LAYOUT CONTRACT — see {@link validateModel}.
 *
 * Returns the union of all problems; `ok` is true only when none are found.
 */
export function validateSource(html: string): ValidationResult {
  let model: DeckModel;
  try {
    model = parseDeck(html);
  } catch (e) {
    return {
      ok: false,
      errors: [
        {
          code: 'parse',
          message: `HTML failed to parse: ${e instanceof Error ? e.message : String(e)}`,
        },
      ],
    };
  }

  const errors = validateModel(model).errors;

  // Round-trip idempotency: serialize(parse(x)) must be a fixed point.
  try {
    const s1 = serializeDeck(model);
    const s2 = serializeDeck(parseDeck(s1));
    if (s1 !== s2) {
      errors.push({
        code: 'round-trip',
        message:
          'Document does not round-trip stably; saving it would alter the markup on reload.',
      });
    }
  } catch (e) {
    errors.push({
      code: 'round-trip',
      message: `Re-parse of serialized output failed: ${e instanceof Error ? e.message : String(e)}`,
    });
  }

  return { ok: errors.length === 0, errors };
}

// ─── Remote (Go) validation — single source of truth, best-effort ───────────

/**
 * Call Lane A's `POST /api/decks/{name}/validate` (spec claude-code-integration "slides validate").
 *
 * Contract assumed (tolerant): the endpoint receives the HTML body and returns
 * JSON `{ ok: boolean, errors?: Array<{ code?, message, eid? }> }`. We map that
 * into a {@link ValidationResult}.
 *
 * Returns `null` (NOT a failure) when the endpoint is UNAVAILABLE — 404 (not yet
 * implemented), a non-2xx, a network error, or non-JSON. A null result means
 * "fall back to the client-side guard as the sole authority" so the editor keeps
 * working offline / before the backend lands the route (spec principles-and-invariants offline-first).
 */
export async function validateRemote(
  name: string,
  source: string,
): Promise<ValidationResult | null> {
  try {
    const res = await fetch(`/api/decks/${encodeURIComponent(name)}/validate`, {
      method: 'POST',
      headers: { 'Content-Type': 'text/html; charset=utf-8' },
      body: source,
    });
    if (!res.ok) return null; // 404 / 5xx → endpoint unavailable, use local guard
    const data: unknown = await res.json();
    return normalizeRemote(data);
  } catch {
    return null; // network error / non-JSON → fall back to local guard
  }
}

/**
 * Normalise an arbitrary JSON payload from the validate endpoint into a
 * {@link ValidationResult}, defensively (the backend may evolve its shape).
 * Exposed for unit testing of the mapping logic without a live server.
 */
export function normalizeRemote(data: unknown): ValidationResult | null {
  if (typeof data !== 'object' || data === null) return null;
  const obj = data as Record<string, unknown>;

  const rawErrors = Array.isArray(obj.errors) ? obj.errors : [];
  const errors: ValidationError[] = rawErrors.map((e) => {
    if (typeof e === 'string') return { code: 'remote', message: e };
    const eo = (e ?? {}) as Record<string, unknown>;
    return {
      code: typeof eo.code === 'string' ? eo.code : 'remote',
      message:
        typeof eo.message === 'string'
          ? eo.message
          : typeof eo.msg === 'string'
            ? eo.msg
            : 'validation error',
      eid: typeof eo.eid === 'string' ? eo.eid : undefined,
    };
  });

  // Prefer an explicit `ok`; otherwise derive it from the presence of errors.
  const ok = typeof obj.ok === 'boolean' ? obj.ok : errors.length === 0;
  return { ok, errors };
}
